packagedb/package_managers.py

```diff
@@ -0,0 +1,635 @@
+    """
+    return version and version.replace("0:", "")
+
+
+class LaunchpadVersionAPI(VersionAPI):
+    """
+    Fetch versions of Ubuntu debian packages from Launchpad
+    """
+
+    package_type = "deb"
+
+    def fetch(self, pkg: str) -> Iterable[PackageVersion]:
+        url = (
+            f"https://api.launchpad.net/1.0/ubuntu/+archive/primary?"
+            f"ws.op=getPublishedSources&source_name={pkg}&exact_match=true"
+        )
+
+        while True:
+            response = get_response(url=url, content_type="json")
+
+            if not response:
+                break
+            entries = response.get("entries")
+            if not entries:
+                break
+
+            for release in entries:
+                source_package_version = release.get("source_package_version")
+                source_package_version = remove_debian_default_epoch(version=source_package_version)
+                date_published = release.get("date_published")
+                release_date = None
+                if date_published and type(date_published) is str:
+                    release_date = dateparser.parse(date_published)
+                if source_package_version:
+                    yield PackageVersion(
+                        value=source_package_version,
+                        release_date=release_date,
+                    )
+            if response.get("next_collection_link"):
+                url = response["next_collection_link"]
+            else:
+                break
+
+
+class PypiVersionAPI(VersionAPI):
+    """
+    Fetch versions of Python pypi packages from the PyPI API.
+    """
+
+    package_type = "pypi"
+
+    def fetch(self, pkg):
+        response = get_response(url=f"https://pypi.org/pypi/{pkg}/json")
+        if not response:
+            # FIXME: raise!
+            return
+
+        releases = response.get("releases") or {}
+        for version, download_items in releases.items():
+            if not download_items:
+                continue
+
+            release_date = self.get_latest_date(download_items)
+            yield PackageVersion(
+                value=version,
+                release_date=release_date,
+            )
+
+    def get_latest_date(self, downloads):
+        """
+        Return the latest date from a list of mapping of PyPI ``downloadss`` or  None.
+
+        The data has this shape:
+        [
+          {
+            ....
+            "upload_time_iso_8601": "2010-12-23T05:14:23.509436Z",
+            "url": "https://files.pythonhosted.org/packages/8f/1f/c20ca80fa5df025cc/Django-1.1.3.tar.gz",
+          },
+          {
+            ....
+            "upload_time_iso_8601": "2010-12-23T05:20:23.509436Z",
+            "url": "https://files.pythonhosted.org/packages/8f/1f/561bddc20ca80fa5df025cc/Django-1.1.3.wheel",
+          },
+        ]
+        """
+        latest_date = None
+        for download in downloads:
```

Comment:
I think its more Pythonic to sort the list of downloads by the upload time, then popping the latest download entry from the list.

```suggestion
        # Sort `downloads` from earliest "upload_time_iso_8601" to latest
        downloads = downloads.sort(key=lambda x: x["upload_time_iso_8601"])
        # Pop the latest download entry from `downloads`
        latest_download = downloads.pop()
        latest_date = latest_download.get("upload_time_iso_8601")
        if latest_date:
            latest_date = dateparser.parse(latest_date)
        return latest_date
```

Reply:
@JonoYang I was trying to keep `package_managers` code untouched and refactor it while migrating to fetchcode.

Reply:
@keshav-space Ah, I see.